### PR TITLE
Feedback to play the game (game logic branch)

### PR DESCRIPTION
**Pull request by thsvr**
Hi, 

Can you please check our code, play with the game and give feedback to us?

In this branch, the game is working and you can play as much as you want :) 

Thank you in advance ;)

Thaís and Sofi

### CODE REVIEW

File: lib/players.rb [hunk 1]
```diff
@@ -0,0 +1,11 @@
+# frozen_string_literal: true
+
+class Player
+  attr_accessor :name, :element, :score
```
Comment (mcrd25) [inline lib/players.rb:4]: Ought not to use `attr_accessor` when possible.
you should rather use `attr_reader` or `attr_writer` respectively.
Using `attr_accessor` allows for rewriting of `:name`, `:element`and `:score` outside of the class Player.

File: lib/game.rb [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    @cell[0][0] == @cell[1][1] && @cell[1][1] == @cell[2][2]
+  end
+
+  def right_diagonal
+    @cell[2][0] == @cell[1][1] && @cell[1][1] == @cell[0][2]
+  end
+
+  def vertical
+    vertical_one || vertical_two || vertical_three
+  end
```
Comment (mcrd25) [inline lib/game.rb]: is there not a better way to check these winning combinations? 🤔

File: lib/game.rb [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    horizontal_one || horizontal_two || horizontal_three
+  end
+
+  def horizontal_one
+    @cell[0][0] == @cell[0][1] && @cell[0][1] == @cell[0][2]
+  end
+
+  def horizontal_two
+    @cell[1][0] == @cell[1][1] && @cell[1][1] == @cell[1][2]
+  end
```
Comment (mcrd25) [inline lib/game.rb]: [OPTIONAL]
Can you think of a more optimal way to check this?
Currently, you are checking each possibility once. 
Consider if there may be a simpler way.

File: lib/board.rb [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+  def move(element, place)
+    if place <= 3
+      @cell[0][@cell[0].index(place)] = element
+    elsif place <= 6
+      @cell[1][@cell[1].index(place)] = element
+    else
+      @cell[2][@cell[2].index(place)] = element
+    end
+    result = Game.new(@cell)
```
Comment (mcrd25) [inline lib/board.rb:21]: [OPTIONAL]
Do you intend on creating a new Game instance for every move?
Consider changing this in the future.

File: lib/board.rb [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+  end
+
+  def move(element, place)
+    if place <= 3
+      @cell[0][@cell[0].index(place)] = element
+    elsif place <= 6
+      @cell[1][@cell[1].index(place)] = element
+    else
+      @cell[2][@cell[2].index(place)] = element
+    end
```
Comment (mcrd25) [inline lib/board.rb:20]: THis is a good candidate for a private method.

Should follow the single responsibility rule for methods

File: lib/game.rb [hunk 6]
```diff
@@ -0,0 +1,59 @@
+# frozen_string_literal: true
+
+class Game
+  attr_accessor :cell
```
Comment (mcrd25) [inline lib/game.rb:4]: See the previous comment about `attr_accessor` and address

**Review by mcrd25 — CHANGES_REQUESTED** [review]
# Changes Required, but great Work 💪 

Hi @thsvr and @SofiBretz 👋,

Wow, great job. I only have some minor things for you to address.

See review comments for more details

### Required Changes
- all requested changes in comments **NOT** marked a [OPTIONAL]

### Additional Suggested Changes [OPTIONAL]
*Please note that these changes are not required*
- All review Comments with requested changes marked as [OPTIONAL]
- I strongly recommend you fix the error where when the user is prompted to replay. When they reply to with anything other than no the game is reset. However, this is not good as yes may not have been the response intended and as it may even have been a typo. Users should be prompted again if the answer is not clear or yes or no.
See image below for better context:

Screen Shot 2019-10-10 at 11 19 56 PM

Please address minor changes requested **NOT** marked as [OPTIONAL] and submit another review. Really excellent job so far 👏 👍 
- Your stickler works fine
- the game is played well
- nothing breaks too badly
- the script file (main.rb) is good!


***Please makes changes and submit another review request.
If you have any questions, let me know here or on slack and I will be glad to clarify with you. You can also reach me and connect with me on my social media or email.***

*Happy coding 😄,*
**[--Maya](https://github.com/mcrd25)**
[GitHub](https://github.com/mcrd25) [Twitter](https://twitter.com/mcrd25) [linkedin](https://www.linkedin.com/in/mayadouglas/) [GitLab](https://gitlab.com/mcrd25)

---
*P.S. all requested changes not marked by [OPTIONAL] or are not positive comments are **REQUIRED***

**Comment by SofiBretz** [comment]
### Summary of changes made

- Rubocop ready, all checks have passed.

- attr_reader and attr_writer fixed.

- Grammar errors corrected and small details also.

We made all the requiments asked by Maya, can you please check if our project is able to be approved?  

Thanks! :)

Sofi and Thaís.

File: lib/players.rb [hunk 7]
```diff
@@ -0,0 +1,12 @@
+# frozen_string_literal: true
+
+class Player
+  attr_accessor :name, :element
+  attr_reader :score
+
+  def initialize(name)
+    @name = name
+    @element = nil
+    @score = 0
```
Comment (Ispirett) [inline lib/players.rb:9]: /////////
X | O | X
/////////
O | X | O
/////////
X | 8 | 9
/////////
name, you won!!!
Traceback (most recent call last):
        4: from bin/main.rb:70:in `<main>'
        3: from bin/main.rb:70:in `loop'
        2: from bin/main.rb:75:in `block in <main>'
        1: from bin/main.rb:75:in `loop'
bin/main.rb:87:in `block (2 levels) in <main>': undefined method `score=' for #<Player:0x0000563423acc710 @name="name", @element="X", @score=0> (NoMethodError)

Got this error which is caused but not allowing the score to be set,  you need to add it to attr_accessor

File: lib/game.rb (hunk 3 above)
Comment (Ispirett) [reply lib/game.rb]: As @mcrd25 said this code is to hardcoded. 
You can use an approach like this

  def horizontal
    hor = [
      [@cell[0][0], @cell[0][1], @cell[0][2]],
      [@cell[1][0], @cell[1][1], @cell[1][2]],
      [@cell[2][0], @cell[2][1], @cell[2][1]]
    ]
  end

```
def check_win?(array)
    array.each {|cell| cell.all?(player.element || player2.element) }
end
```
This is untested but its just to give an example of how you can approach this.

**Review by Ispirett — CHANGES_REQUESTED** [review]
Really great Work @SofiBretz @thsvr  with the project and changes made :+1:  :100:  like how you used the sleep method looks cool, and the display messages. 

However there some bugs :bug:  and minor adjustments to be made.

You should also move this ` ** => board = Board.new**` out of loop do 
everytime you loop reloops its going create a new board

**Comment by SofiBretz** [comment]
### Summary of changes made

- Rubocop ready, all checks have passed.

- Score accessor fixed.

- Suggestion "As @mcrd25 said this code is to hardcoded. You can use an approach like this", we test this suggestion and didn't work, broked our code. We tried with our code and works, thank you for your suggestion. 

We made all the requirements asked by Maya and Ispirett, can you please check if our project is able to be approved?

Thanks! :)

Sofi and Thaís.

File: lib/game.rb [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+    horizontal_one || horizontal_two || horizontal_three
+  end
+
+  def horizontal_one
+    @cell[0][0] == @cell[0][1] && @cell[0][1] == @cell[0][2]
+  end
+
+  def horizontal_two
+    @cell[1][0] == @cell[1][1] && @cell[1][1] == @cell[1][2]
+  end
```
Comment (Ispirett) [inline lib/game.rb:56]: This example was just to ignite some ideas and will not work as-is. Currently, your code breaks the Dry principle (don't repeat your self). There are more efficient solutions with less repetition.  Hint!! you can use ranges also eg: [1..3] 

note this referring to the methods below private access modifier

**Review by Ispirett — CHANGES_REQUESTED** [review]
Really good work with changes and refactoring @SofiBretz @thsvr :+1:  :1st_place_medal: 
The only thing is this small optimization. 
I know you can do it based on the great work you have already done !!!
**almost there.**

**Comment by SofiBretz** [comment]
### Summary of changes made

- Rubocop ready, all checks have passed.

- lib/game.rb methods fixed, now it's faster than before.

- Suggestion "As @Ispirett  "There are more efficient solutions with less repetition..." We solve it, thanks for the suggestion :)

We made all the requirements asked by Maya, Ispirett and Bola, can you please check if our project is able to be approved?

Thanks! :)

Sofi and Thaís.

**Review by whiz25 — APPROVED** [review]
## Nice work :clap: 
You have done so much with this project and therefore it's a :+1:  from me. Please merge and proceed.

#### Martin Afani

**Pull request closed without merging**